server/controllers/errorController.js

```diff
@@ -0,0 +1,168 @@
+    console.log(errorList[0].length, errorList[1].length, errorList[2].length)
+    //Pop the last row since it is empty
+    errorList.pop();
+    //Shift the first row since it is the headers
+    res.locals.errorList = errorList;
+    return next();
+  } catch (err) {
+    return next({
+      log: `errorController.queryErrors: ERROR: ${err}`,
+      message: {
+        err: 'Error occurred while querying errors. Check server logs for more information.',
+      }
+    })
+  }
+};
+
+//Store an array of the index of where each header in HEADERS array begins as well as the array length
+function headersIndex(array) {
+  let arr = [];
+  for (let i = 0; i < HEADERS.length; i++) {
+    arr.push(array.indexOf(HEADERS[i]));
+  }
+  //arr.push(array.length);
+  console.log('headersIndex', arr);
+  return arr;
+}
+
+//Converts ErrorList array of strings to nested array
+function errorArrayConverter(array) {
+  let arr = [];
+  let headersIndexArray = headersIndex(array[0])
+  for (let i = 0; i < array.length; i++) {
+    let subArray = [];
+    for (let j = 0; j < headersIndexArray.length - 1; j++) {
+      subArray.push(array[i].slice(headersIndexArray[j], headersIndexArray[j + 1]).trim())
+      if (j === headersIndexArray.length - 2) {
+        subArray.push(array[i].slice(headersIndexArray[j + 1], array[i].length).trim())
+      }
+    }
+    arr.push(subArray);
+  }
+  return arr;
+}
+
+errorController.formatErrors = async (req, res, next) => {
+  try {
+    let headersIndexArray = headersIndex(res.locals.errorList[0])
+    let formattedError = errorArrayConverter(res.locals.errorList);
+    //const { errorList } = req.locals;
+    //const formattedErrors = [];
+    res.locals.formattedError = formattedError;
+    return next();
+  } catch (err) {
+    return next({
+      log: `errorController.formatErrors: ERROR: ${err}`,
+      message: {
+        err: 'Error occurred while formatting errors. Check server logs for more information.',
+      }
+    })
+  }
+};
+
+
+/*
+
+error = {
+  namespace,
+  lastSeen,
+  type,
+  reason,
+  object,
+}
+
+*/
+
+errorController.saveErrors = async (req, res, next) => {
+  try {
+    let formattedError = res.locals.formattedError;
+    // 1. Iterate through all our errors
+    // 2. save each instance of error; this will by async work
+    // 3. once all errors are saved, we want to find and return up all errors
+    // 4. save those errors to res.locals
+    // 5. hit that next piece of sweet sweet middleware
+    for (let i = 0; i < formattedError.length; i++) {
+      const obj = {
+        namespace: formattedError[i][0],
+        lastSeen: formattedError[i][1],
+        type: formattedError[i][2],
+        reason: formattedError[i][3],
+        object: formattedError[i][4],
+        message: formattedError[i][5],
+      }
+      Error.create(obj)
+    }
+    //res.locals.error
+
+    // const errors = Errors.findBy({});
+    // console.log(errors);
+    return next();
```

Comment:
How are we controlling for async work here? It seems like we would move to the next piece of middleware once we're done iterating through formattedError (is this one error or many??) This seems like potentially a good place to use `Promise.all`— @JefZheng @Hydroelectric29 any thoughts on that?